verl/experimental/agent_loop/tool_agent_loop.py

```
@@ -438,6 +438,18 @@ async def _call_tool(
         try:
             # TODO: append malformed tool_call to the prompt: invalid function name or arguments
             tool_name = tool_call.name
+
+            # Validate tool exists before attempting to execute
+            if tool_name not in self.tools:
+                available_tools = ", ".join(self.tools.keys()) if self.tools else "none"
```

We can raise an exception here, and handle exception uniformly.

---

replied in the issue https://github.com/volcengine/verl/issues/4124#issuecomment-3568879862